### DISCUSSION TITLE
New Top panel

### DISCUSSION THREAD

**Discussion by YathindraNikshipth** [eclipse-theia/theia#13934] [Q&A]
I want to create a new top panels. Instead of one i want to create three to four top panels with varied widths with some spacing between each panels. Is there a way to do it?

**Comment by msujew** [comment]
Hey @YathindraNikshipth,

it probably makes sense to take a look at our [ApplicationShellWithToolbarOverride](https://github.com/eclipse-theia/theia/blob/9e9a9f11d6f4654eb36294aed64d20a4a776da52/packages/toolbar/src/browser/application-shell-with-toolbar-override.ts#L71-L93) where we do exactly this kind of stuff. Essentially, you need to override the `createLayout` method of the `ApplicationShell` in a way that lets you add new panels into the layout (ideally without removing existing panels). The `ApplicationShellWithToolbarOverride` in particular just adds a single panel right under the `topPanel` to add a toolbar to the application.

**Comment by YathindraNikshipth** [comment]
Hi, @msujew  thanks for the reply, I will look into the class u mentioned.

**Comment by BarryDrez** [comment]
Hi @msujew,
I have extended ApplicationShell over a year ago with Version 1.40 of Theia.  I am trying to upgrade to 1.51.0, but my ApplicationShell is not working anymore.
I have overridden the addWidget function.  In 1.40, my function would be called by other classes.  This no longer happens in 1.51.  Now the calls go directly to ApplicationShell.  Any ideas on how I can handle this?

**Reply by msujew** [reply]
Is your class bound correctly in inversify? Maybe you have added the `@theia/toolbar` package recently, which has its own application shell?

**Reply by BarryDrez** [reply]
@msujew You are correct.  I had it bound as follows:
    bind(MyApplicationShell).toSelf().inSingletonScope();
    rebind(ApplicationShell).to(MyApplicationShell).inSingletonScope();

This works under 1.40. It compiles under 1.51, but does not bind correctly.

So I changed it to:
    bind(MyApplicationShell).toSelf().inSingletonScope();
    rebind(ApplicationShell).toService(MyApplicationShell);

and now it works.

Thank you!
